### PR TITLE
Improve useProgressRing

### PR DESCRIPTION
**Pull request by mparvazi**

### CODE REVIEW

File: src/js/components/useProgressRing.ts [hunk 1]
```diff
@@ -33,19 +34,38 @@ export interface ProgressRingReturn {
...
-    const circle = element.querySelector<SVGCircleElement>(ProgressRingMap.checkout.circle);
+export const TextType = {
+  enum: 'enum',
+  percent: 'percent',
+};
+
+export interface ProgressRingOptions {
+  steps: number;
+  text: keyof typeof TextType;
+}
```
Comment (NeOMakinG) [inline src/js/components/useProgressRing.ts:45]:
```suggestion
export enum TextType {
  enum: 'enum',
  percent: 'percent',
};

export interface ProgressRingOptions {
  steps: number;
  text: TextType;
}
```

Does it works like this?

File: src/js/components/useProgressRing.ts [hunk 2]
```diff
@@ -33,19 +34,38 @@ export interface ProgressRingReturn {
   error?: Error
```
Comment (NeOMakinG) [inline src/js/components/useProgressRing.ts:34]:
```suggestion
  error?: Error
  progressElement: HTMLElement
```

File: src/js/components/useProgressRing.ts [hunk 3]
```diff
@@ -33,19 +34,38 @@ export interface ProgressRingReturn {
...
         circle.style.strokeDashoffset = offset.toString();
+        circle.dataset.percent = String(percent);
+
+        if (progressText) {
+          const text = (options.text === TextType.enum) ? `${index} / ${options.steps}` : `${percent}%`;
+          progressText.innerHTML = text;
+        }
       };
 
       return {
```
Comment (NeOMakinG) [inline src/js/components/useProgressRing.ts:71]:
```suggestion
      return {
        progressElement,
```

File: templates/components/progress-circle.tpl [hunk 4]
```diff
@@ -48,14 +46,12 @@
...
     stroke="currentColor"
     stroke-width="{$stroke}"
-    data-percent="50"
+    data-percent="0"
     fill="transparent"
     r="{$radius}"
     cx="{$size / 2}"
     cy="{$size / 2}"
-    style="stroke-dasharray: {$circumference}, {$circumference}; stroke-dashoffset: {$dashoffset};"
+    style="stroke-dasharray: {$circumference}, {$circumference}; stroke-dashoffset: {$circumference};"
```
Comment (NeOMakinG) [inline templates/components/progress-circle.tpl:54]: Are you sure? It's working fine?

File: src/js/components/useProgressRing.ts (hunk 1 above)
Comment (mparvazi) [reply src/js/components/useProgressRing.ts:45]: In modern TypeScript, you may not need an enum when an object with as const could suffice:
https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums

but here i forget to add `as const`  to the TextType object!

File: templates/components/progress-circle.tpl (hunk 4 above)
Comment (mparvazi) [reply templates/components/progress-circle.tpl:54]: In initial state (0%) the circle should be gray.
{$dashoffset = $circumference - (($percent / 100 ) * $circumference)}
so when $percent is 0 => $dashoffset = $circumference

File: src/js/components/useProgressRing.ts (hunk 1 above)
Comment (NeOMakinG) [reply src/js/components/useProgressRing.ts:45]: But using an enum for such a thing looks like a good practice on different projects

Comment (mparvazi) [reply src/js/components/useProgressRing.ts:45]: I've tested, there is a problem with eslint when using enum here.
I think it's a bug in eslint:
https://stackoverflow.com/questions/63961803/eslint-says-all-enums-in-typescript-app-are-already-declared-in-the-upper-scope

Comment (NeOMakinG) [reply src/js/components/useProgressRing.ts]: Isn't it related to the enum keyword that should be named differently?

Comment (mparvazi) [reply src/js/components/useProgressRing.ts]: Using object with `as const` is safer:
Screenshot from 2022-03-29 20-35-33

**Comment by NeOMakinG** [comment]
@mparvazi what's up with this?

**Comment by mparvazi** [comment]
Can not use enum with eslint.

**Comment by NeOMakinG** [comment]
> Can not use enum with eslint.

I guess it's not related, it was because it was already defined?

Do you want me to get a look and open this PR again? Why did you close it?

**Comment by mparvazi** [comment]
@NeOMakinG 

Try to use an enum inside a component and test with eslint.

**Comment by NeOMakinG** [comment]
@mparvazi Fixed it :)

I guess we're near the merge right?

**Comment by mparvazi** [comment]
no-shadow: true

then:
```
var a = 3;
function b() {
    var a = 10;
}
```
In this case, the variable a inside of b() is shadowing the variable a in the global scope. 
no problem about this?

**Comment by mparvazi** [comment]
OK, so let me check again.

**Comment by mparvazi** [comment]
@NeOMakinG 
Ready to merge.

**Comment by NeOMakinG** [comment]
@mparvazi let me test it tomorrow and it looks pretty fine to me now :)

**Review by NeOMakinG — APPROVED** [review]
The implementation looks fine to me, also tested it and it works fine

As we discussed in private weeks ago, we should think about testing `useProgress` and `useAlert` if it's possible

**Pull request merged**